### PR TITLE
Get point cloud as a buffer

### LINKED ISSUE #2283: Point cloud performance improvement
**Describe the Bug**
Method `lidar.getPointCloud()` (Python) is a few times slower than `wb_get_point_cloud()` (C).

See conversation:
https://github.com/cyberbotics/webots_ros2/pull/148#issuecomment-702001390

**Steps to Reproduce**
1. Run the following world in Webots: https://github.com/cyberbotics/webots_ros2/blob/b4d79ee9259ead364dbc7ee50ad7115e39ba5cb6/webots_ros2_universal_robot/worlds/universal_robot_lidar.wbt
2. Use the Python controller (see **Additional context**) and check the real-time factor.
2. Use the C controller (see **Additional context**) and check the real-time factor.

**Expected behavior**
Method `lidar.getPointCloud()` (Python) should be nearly the speed of `wb_get_point_cloud()` (C).

**Additional context**
<details>
<summary>Example C controller</summary>

```c
#include <stdio.h>
#include <webots/robot.h>
#include <webots/lidar.h>


#define TIME_STEP 32


int main(int argc, char **argv) {
  wb_robot_init();
  int device = wb_robot_get_device("Velodyne VLP-16");
  wb_lidar_enable(device, TIME_STEP);
  wb_lidar_enable_point_cloud(device);
  
  while (wb_robot_step(TIME_STEP) != -1) {
    WbLidarPoint* points = wb_lidar_get_point_cloud(device);
  };

  wb_robot_cleanup();

  return 0;
}
```

</details>

<details>
<summary>Example Python controller</summary>

```Python
from controller import Robot


robot = Robot()
timestep = 32
lidar = robot.getLidar('Velodyne VLP-16')
lidar.enable(timestep)
lidar.enablePointCloud()

while robot.step(timestep) != -1:
    vals = lidar.getPointCloud()
    sum_val = sum(val.x for val in vals)
    print(sum_val)
```

</details>

### PR DESCRIPTION
**Pull request by lukicdarkoo**
**Description**
Since I couldn't find a solution to improve the performance of `getPointCloud()` without changing the API (see #2296) this introduces a new argument `dtype` which defines data type.

See conversation:
https://github.com/cyberbotics/webots_ros2/pull/148#issuecomment-702001390

**Related Issues**
Closes #2283

**Tasks**
 - [x] Add to changelog
 - [x] Add documentation

**API Proposal**

`getPointCloud([dtype])`
- `dtype` can be ['list', 'buffer']

**Additional context**
- If `dtype` == 'list'
  - Creating a list is 50-60% faster than the previous implementation
- If `dtype` == 'buffer'
  - Performances are almost the same as for C API
  - Data can be passed directly to [`PointCloud2`](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/PointCloud2.html) (see [example](https://gist.github.com/pgorczak/5c717baa44479fa064eb8d33ea4587e0))
  - It can be converted to numpy array using `data = np.frombuffer(data, dtype=np.uint8)` without overhead
- This change doesn't break the API
- It can scale to other parts of the API without breaking it
- We can continue adding new data types

----

**Personal notes:**

- Constructor for `array.array` (`newarrayobject`):
https://github.com/python/cpython/blob/e51dd9dad6590bf3a940723fbbaaf4f64a3c9228/Modules/arraymodule.c#L570

- Creates `array.array` object by calling `array_array_frombytes`:
https://github.com/python/cpython/blob/e51dd9dad6590bf3a940723fbbaaf4f64a3c9228/Modules/arraymodule.c#L2663

- Bytes to `array.array` is only doing `memcpy()` (and some checks):
https://github.com/python/cpython/blob/e51dd9dad6590bf3a940723fbbaaf4f64a3c9228/Modules/arraymodule.c#L1615

- Access to memory without copying (only get):
```Python
mv = memoryview(data)
mv[0:100]
```

- Pointer to the memory of `array.array`:
```C
op->ob_item
```

- The internal buffer cannot be alternated through [`PyObject_GetBuffer`](https://docs.python.org/3/c-api/buffer.html#c.PyObject_GetBuffer) as `view->buf` is assigned on the call:
https://github.com/python/cpython/blob/e51dd9dad6590bf3a940723fbbaaf4f64a3c9228/Modules/arraymodule.c#L2523

-

### CODE REVIEW

**Comment by lukicdarkoo** [comment]
I found a way to pass `bytearray` to ROS2 without mimicking `array.array`:
https://github.com/cyberbotics/webots_ros2/blob/4a02a97659ab84c8bd01e72eb1ac9d7ed7e4e6b0/webots_ros2_core/webots_ros2_core/devices/lidar_device.py#L114

It is actually a straightforward solution, but I thought it would not work as the data type is not the same.

File: resources/languages/python/controller.i [hunk 1]
```diff
@@ -557,17 +557,40 @@ class AnsiCodes(object):
     return point[index];
   }
 
+  PyObject* __getPointCloudBuffer() const {
+    const char * points = (const char *)$self->getPointCloud();
```
Comment (omichel) [inline resources/languages/python/controller.i:561]:
```suggestion
    const char *points = (const char *)$self->getPointCloud();
```

File: resources/languages/python/controller.i [hunk 2]
```diff
@@ -557,17 +557,40 @@ class AnsiCodes(object):
...
+    const int size = $self->getNumberOfPoints() * sizeof(WbLidarPoint);
+    return PyBytes_FromStringAndSize(points, size);
+  }
+
+  PyObject* __getPointCloudList() const {
+    const WbLidarPoint* rawPoints = $self->getPointCloud();
```
Comment (omichel) [inline resources/languages/python/controller.i:567]:
```suggestion
    const WbLidarPoint *rawPoints = $self->getPointCloud();
```

File: resources/languages/python/controller.i [hunk 3]
```diff
@@ -557,17 +557,40 @@ class AnsiCodes(object):
     return point[index];
   }
 
+  PyObject* __getPointCloudBuffer() const {
```
Comment (omichel) [inline resources/languages/python/controller.i:560]:
```suggestion
  PyObject *__getPointCloudBuffer() const {
```

**Review by omichel — CHANGES_REQUESTED** [review]
That solution looks great and simple!
I only have a doubt about the `dtype` naming (I saw this is used in numpy, but usually, we prefer a more standard naming scheme). I would suggest `data_type` instead.

**Comment by lukicdarkoo** [comment]
> That solution looks great and simple!
> I only have a doubt about the `dtype` naming (I saw this is used in numpy, but usually, we prefer a more standard naming scheme). I would suggest `data_type` instead.

Yes, `dtype` comes from numpy, but `data_type` is prettier, I will change it.

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**